### PR TITLE
arch: xtensa: save FPU register in context switching

### PR DESCRIPTION
**Pull request by LucasTambor**
Save FP user register and FP register file during context switch.

This commit enables shared FP registers mode using CONFIG_FPU_SHARING.

Since there is no lazy stacking, the FPU registers will be saved regardless
of whether floating point calculations are performed in the threads when
CONFIG_FPU_SHARING is enabled. This require 72 additional bytes in the stack
memory.

Signed-off-by: Lucas Tamborrino [lucas.tamborrino@espressif.com](mailto:lucas.tamborrino@espressif.com)

### CODE REVIEW

File: arch/xtensa/core/xtensa-asm2-util.S [hunk 1]
```diff
@@ -153,6 +153,10 @@ _restore_context:
 	l32i a0, a1, BSA_PS_OFF
 	wsr a0, ZSR_EPS
 
+#if XCHAL_HAVE_FP && defined(CONFIG_CPU_HAS_FPU) && defined(CONFIG_FPU_SHARING)
```
Comment (andyross) [inline arch/xtensa/core/xtensa-asm2-util.S:156]: Better to use just one symbol here and not double-test.  Just use a BUILD_ASSERT() to verify that the kconfig and the core-isa.h symbol match.

File: arch/xtensa/include/xtensa-asm2-context.h [hunk 2]
```diff
@@ -127,4 +134,15 @@
...
 
+#if XCHAL_HAVE_FP && defined(CONFIG_CPU_HAS_FPU) && defined(CONFIG_FPU_SHARING)
+#define BSA_FPU_OFF \
+			(BASE_SAVE_AREA_SIZE - \
+			 (BASE_SAVE_AREA_SIZE_COMMON + \
+			  BASE_SAVE_AREA_SIZE_LOOPS + \
+			  BASE_SAVE_AREA_SIZE_EXCCAUSE + \
+			  BASE_SAVE_AREA_SIZE_SCOMPARE + \
+			  BASE_SAVE_AREA_SIZE_THREADPTR + \
+			  BASE_SAVE_AREA_SIZE_FPU))
```
Comment (andyross) [inline arch/xtensa/include/xtensa-asm2-context.h:145]: Every time I see this code I cringe a little.  It hasn't scaled well, but this is indeed the scheme we've picked so there's nothing wrong with extending it in the intended way.  Could do with a rewrite though.

**Review by andyross — DISMISSED** [review]
This looks clean to me.  One style nitpick on preprocessor configuration management, but nothing that should block merge.  It would be nice to get it working in a lazy-evaluated way.  This is going to add a ton of cycles to interrupt entry needlessly.

File: arch/xtensa/include/xtensa-asm2-context.h [hunk 3]
```diff
@@ -85,12 +85,19 @@
 #define BASE_SAVE_AREA_SIZE_THREADPTR	0
 #endif
 
+#if XCHAL_HAVE_FP && defined(CONFIG_CPU_HAS_FPU) && defined(CONFIG_FPU_SHARING)
+#define BASE_SAVE_AREA_SIZE_FPU		18 * 4
```
Comment (uLipe) [inline arch/xtensa/include/xtensa-asm2-context.h:89]: Nit: Why not put the final value directly here?

**Review by uLipe — DISMISSED** [review]
LGTM!

File: arch/xtensa/core/xtensa-asm2-util.S (hunk 1 above)
Comment (LucasTambor) [reply arch/xtensa/core/xtensa-asm2-util.S:156]: @andyross , thanks for the nitpicks. I'm having some trouble using the BUILD_ASSERT() macro on the xtensa-asm2-util.S or xtensa-asm2-s.h. 

```
#if defined(CONFIG_FPU_SHARING)
BUILD_ASSERT(XCHAL_HAVE_FP & CONFIG_CPU_HAS_FPU)
#endif
```
It keeps failing to compile with the error:
```
zephyr/arch/xtensa/include/xtensa-asm2-s.h: Assembler messages:
zephyr/arch/xtensa/include/xtensa-asm2-s.h:13: Error: unknown opcode or format name 'build_assert'
```

Any tip on using it?

File: arch/xtensa/include/xtensa-asm2-context.h (hunk 3 above)
Comment (LucasTambor) [reply arch/xtensa/include/xtensa-asm2-context.h:89]: I find it more readable than plain "72". I think it becomes clearer that an additional 18 x 32bit space on the stack is needed. But I can change it if you disagree, no problem at all.

Comment (dcpleung) [reply arch/xtensa/include/xtensa-asm2-context.h:89]: Please put parentheses around this.

File: arch/xtensa/core/xtensa-asm2-util.S (hunk 1 above)
Comment (LucasTambor) [reply arch/xtensa/core/xtensa-asm2-util.S:156]: Ahh, thanks @dcpleung !

**Review by sylvioalves — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by LucasTambor** [comment]
FYI the compliance check error seems to be a false positive, it is misinterpreting the assembly instruction as a declaration :)

**Review by andyross — DISMISSED** [review]
refresh +1

**Review by dcpleung — APPROVED** [review]

**Review by dcpleung — CHANGES_REQUESTED** [review]
Please rebase to remove the merge commit.

**Review by dcpleung — DISMISSED** [review]
Compliance check is a false positive.

**Review by uLipe — DISMISSED** [review]

**Review by sylvioalves — DISMISSED** [review]

**Comment by sylvioalves** [comment]
@dcpleung @ceolin, does this CI false positive check need to be handled somehow?

**Review by dcpleung — DISMISSED** [review]
Compliance check doesn't do a good job on assembly. We can ignore that warning.

File: tests/kernel/fpu_sharing/generic/src/float_context.h [hunk 4]
```diff
@@ -150,6 +150,19 @@ struct fp_non_volatile_register_set {
...
+struct fp_volatile_register_set {
+	/* No volatile floating point registers */
+};
+
+struct fp_non_volatile_register_set {
+	uint32_t reg[18];     /* FR register file consists of 18 registers of 32 bits */
+};
+
+#define SIZEOF_FP_VOLATILE_REGISTER_SET sizeof(struct fp_volatile_register_set)
+#define SIZEOF_FP_NON_VOLATILE_REGISTER_SET 0
```
Comment (ceolin) [inline tests/kernel/fpu_sharing/generic/src/float_context.h:164]: These defintions are inverted. `SIZEOF_FP_NON_VOLATILE_REGISTER_SET` should be `sizeof(struct fp_non_volatile_register_set)` and the volatile 0.

**Comment by LucasTambor** [comment]
@ceolin thanks for pointing the issue out. Can you review it again, please?

**Review by dcpleung — APPROVED** [review]

**Review by uLipe — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Comment by sylvioalves** [comment]
Hi, this PR is ready to me merged.

**Pull request merged**